backup: Improve wording for --one-file-system description

rawtaz opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

Improve wording for --one-file-system description to make it clearer that it also applies to mountpoints and subvolumes.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Closes #2109.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [ ] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

rawtaz committed:
backup: Improve wording for --one-file-system description

rawtaz commented:
I'm wondering if this is too specific, e.g. can there be subvolumes that *will* be crossed into by restic? If so, the description will be misguiding for those cases. Otherwise I think it should be fine even if it doesn't explicitly mention BTRFS.

MichaelEischer commented:
What happens when using bind mounts? These probably still count as the same filesystem.

MichaelEischer commented:
I just tested bind mounts, as I presumed the `--one-filesystem` option crosses bind mounts within the same filesystem. Maybe just state that the option `exclude[s] other file systems and subvolumes`?

rawtaz commented:
Does anyone have an ZFS setup they can test with to see how traversing is done with that type of filesystem?

MichaelEischer commented on doc/040_backup.rst:
```diff
@@ -216,7 +216,7 @@ On most Unixy shells, you can either quote or use backslashes. For example:
 to only backup files from the file systems the initially specified files
 or directories reside on. In other words, it will prevent restic from crossing
-filesystem boundaries when performing a backup.
+filesystem boundaries, mountpoints and subvolumes when performing a backup.
```
I'd prefer to just explicitly add `subvolumes` to the previously mentioned `filesystem`. mountpoints on their own are only a necessary, but no sufficient condition. For example bind mounts within a filesystem are crossed even with the `--one-file-system` option.

rawtaz commented:
> Does anyone have an ZFS setup they can test with to see how traversing is done with that type of filesystem?

I just tested this, and `--one-file-system` does not have any effect on a ZFS filesystem being mounted, despite ZFS calling it a "filesystem":

~~~
# pwd
/root


# ./restic -r temp backup /data/foo/apa.txt
enter password for repository:
repository 945fd9d4 opened successfully, password is correct
created new cache in /root/.cache/restic
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            2 new,     0 changed,     0 unmodified
Added to the repo: 1.073 KiB

processed 1 files, 5 B in 0:00
snapshot 9a06d7c1 saved


# ./restic -r temp backup --one-file-system /data/foo/apa.txt
enter password for repository:
repository 945fd9d4 opened successfully, password is correct
using parent snapshot 9a06d7c1

Files:           0 new,     0 changed,     1 unmodified
Dirs:            0 new,     0 changed,     2 unmodified
Added to the repo: 0 B

processed 1 files, 5 B in 0:00
snapshot f44eb4da saved


# vi /data/foo/apa.txt


# ./restic -r temp backup --one-file-system /data/foo/apa.txt
enter password for repository:
repository 945fd9d4 opened successfully, password is correct
using parent snapshot f44eb4da

Files:           0 new,     1 changed,     0 unmodified
Dirs:            0 new,     2 changed,     0 unmodified
Added to the repo: 1.085 KiB

processed 1 files, 16 B in 0:00
snapshot 1cc4af11 saved


# vi /data/foo/apa.txt


# ./restic -r temp backup -x /data/foo/apa.txt
enter password for repository:
repository 945fd9d4 opened successfully, password is correct
using parent snapshot 1cc4af11

Files:           0 new,     1 changed,     0 unmodified
Dirs:            0 new,     2 changed,     0 unmodified
Added to the repo: 1.096 KiB

processed 1 files, 27 B in 0:00
snapshot 8be5ba5b saved
~~~

rawtaz commented:
So, is the situation with `--one-file-system` as follows?:

1. Regular filesystem-on-other-disk/device are not traversed into.
1. Bind mounts are traveresed into.
1. ZFS filesystems, even though they are on a completely separate pool on separate disks, are traversed into (this is quite contrary to # 1 above).
1. BTRFS need to be tested as well I think.
1. I would expect that LVM volumes that are mounted are *not* traversed into.
1. How it behaves on Windows I haven't tested.

MichaelEischer commented:
Which filesystem is mounted where? If guess `/data/foo` is a mountpoint?

`--one-file-system` doesn't have any effect when just backing up a single file. The current implementation gathers a list of filesystem IDs for all specified backup target paths (`/data/foo/apa.txt` in this case) and then while walking the filesystem ignores all files/folders outside the whitelisted filesystems. 

So a test setup would have to be something like this:
```
/data (on root partition)
/data/foo (on subvolume)
/data/foo/something
```
then run `backup -x /data`. This should normally skip everything in `/data/foo`.

shd2h commented:
I ran through some quick tests for both zfs and btrfs, which hopefully are some help. I set both up with a second volume mounted underneath the filesystem mountpoint.

```
root@ubu2010:~# zfs list
NAME            USED  AVAIL     REFER  MOUNTPOINT
zfspool         190K  9.20G       24K  /zfspool
zfspool/alpha    24K  9.20G       24K  /fs_test/zfs
zfspool/beta     24K  9.20G       24K  /fs_test/zfs/anothervol

root@ubu2010:~# btrfs fi show /fs_test/btrfs
Label: 'antitank'  uuid: 0151d3ba-84b9-4546-b2c2-d1850484322f
        Total devices 1 FS bytes used 216.00KiB
        devid    1 size 10.00GiB used 536.00MiB path /dev/sdb
root@ubu2010:~# btrfs subvol list /fs_test/btrfs
ID 257 gen 9 top level 5 path subvolume1
```


Complete filesystem layout was as below:

```
root@ubu2010:~# ls -lhR /fs_test/btrfs
/fs_test/btrfs:
total 4.0K
-rw-r--r-- 1 root root 2.9K Apr  2 14:35 normal_file
drwxr-xr-x 1 root root   28 Apr  2 14:35 subvolume1

/fs_test/btrfs/subvolume1:
total 4.0K
-rw-r--r-- 1 root root 3.0K Apr  2 14:35 subvolume_file

root@ubu2010:~# ls -lhR /fs_test/zfs
/fs_test/zfs:
total 4.0K
drwxr-xr-x 2 root root    3 Apr  2 14:44 anothervol
-rw-r--r-- 1 root root 2.9K Apr  2 14:35 normal_file

/fs_test/zfs/anothervol:
total 3.5K
-rw-r--r-- 1 root root 3.0K Apr  2 14:35 subvolume_file
```


backing up the parent dir containing both filesystem mountpoints, using the --one-file-system flag, nothing is backed up:

```
root@ubu2010:~# restic backup --one-file-system /fs_test
repository 4739356d opened successfully, password is correct
no parent snapshot found, will read all files

Files:           0 new,     0 changed,     0 unmodified
Dirs:            3 new,     0 changed,     0 unmodified
Added to the repo: 1.023 KiB

processed 0 files, 0 B in 0:00
snapshot d6b1b075 saved
root@ubu2010:~# restic ls d6b1b075
repository 4739356d opened successfully, password is correct
snapshot d6b1b075 of [/fs_test] filtered by [] at 2021-04-02 14:55:24.212439704 +0000 UTC):
/fs_test
/fs_test/btrfs
/fs_test/zfs
```


backing up the btrfs mountpoint, the contents of the subvolume are excluded:

```
root@ubu2010:~# restic backup --one-file-system /fs_test/btrfs
repository 4739356d opened successfully, password is correct
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            3 new,     0 changed,     0 unmodified
Added to the repo: 4.224 KiB

processed 1 files, 2.857 KiB in 0:00
snapshot 9ae34f42 saved
root@ubu2010:~# restic ls 9ae34f42
repository 4739356d opened successfully, password is correct
snapshot 9ae34f42 of [/fs_test/btrfs] filtered by [] at 2021-04-02 14:55:29.844439118 +0000 UTC):
/fs_test
/fs_test/btrfs
/fs_test/btrfs/normal_file
/fs_test/btrfs/subvolume1
```


doing the same for zfs, the contents of the second "nested" volume are excluded:

```
root@ubu2010:~# restic backup --one-file-system /fs_test/zfs
repository 4739356d opened successfully, password is correct
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            3 new,     0 changed,     0 unmodified
Added to the repo: 1.360 KiB

processed 1 files, 2.857 KiB in 0:00
snapshot ad22eaff saved
root@ubu2010:~# restic ls ad22eaff
repository 4739356d opened successfully, password is correct
snapshot ad22eaff of [/fs_test/zfs] filtered by [] at 2021-04-02 14:55:32.980258298 +0000 UTC):
/fs_test
/fs_test/zfs
/fs_test/zfs/anothervol
/fs_test/zfs/normal_file
```

MichaelEischer commented:
@shd2h Thanks for the tests.

So this confirms that subvolumes are just treated like filesystems and thus `--one-file-system` does not traverse mountpoints to other filesystems or subvolumes. 

@rawtaz The documentation should then read:
`[...] it will prevent restic from crossing filesystem boundaries and subvolumes when performing a backup [...]`

MichaelEischer reviewed (approved):
LGTM.
